docs/proposals/20220712-cluster-api-addon-orchestration.md

```diff
@@ -0,0 +1,437 @@
+
+2. Advanced orchestration will build on the foundation from the first iteration and will explore and add scenarios like configuring add-ons to be upgraded before the Cluster upgrade starts. This work will leverage the lifecycle hooks introduced in Cluster API v1.2.
+
+3. Add-on dependencies can be explored after the orchestration becomes more mature. This iteration can consider use cases where there are dependencies between add-ons and eventually introduce capabilities to account for that in add-on orchestration. For example, add-ons might need to be installed or ugpraded in a certain order and certain versions of add-ons might be incompatible with each other. The scope of this work is TBD.
+
+## Motivation
+
+ClusterResourceSets are the current solution for managing add-ons in Cluster API. However, ClusterResourceSets have been designed as a temporary measure until a better alternative is available. In particular, an add-on solution in Cluster API should be in line with the [goals of Cluster API](https://cluster-api.sigs.k8s.io/#goals) by aiming “to reuse and integrate existing ecosystem components rather than duplicating their functionality.” These components include other projects in SIG Cluster Lifecycle such as [Cluster Add-ons](https://github.com/kubernetes-sigs/cluster-addons) and well established tools like Helm, kapp, ArgoCD, and Flux. 
+
+This proposal does not intend to solve the issue of add-ons in all of Kubernetes. Rather, it aims to bring add-ons into the Cluster API cluster lifecycle by managing add-ons using a declarative spec and orchestrating existing package management tools.
+
+### Goals
+
+- To design a solution for orchestrating Cluster add-ons.
+- To leverage existing package management tools such as Helm for all the foundational capabilities of add-on management, i.e. add-on packages/repository, templating/configuration, add-on creation, upgrade and deletion etc. 
+- To make add-on management in Cluster API modular and pluggable, and to make it simple for developers to build a Cluster API Add-on Provider based on any package management tool, just like with infrastructure and bootstrap providers.
+
+### Non goals
+
+- To implement a full fledged package management tool in Cluster API; there are already several awesome package management tools in the ecosystem, and CAPI should not reinvent the wheel. 
+- To provide a mechanism for altering, customizing, or dealing with single Kubernetes resources defining a Cluster add-on, i.e. Deployments, Services, ServiceAccounts. Cluster API should treat add-ons as opaque components and delegate all the operations impacting add-on internals to the package management tool.
+- To expect users to use a specific package management tool.
+- To implement a solution for installing add-ons on the management cluster itself.
+
+### Future work
+
+- Handle the upgrade of Cluster add-ons when the workload Cluster is upgraded to a new Kubernetes version.
```

Comment:
> when the workload Cluster is upgraded to a new Kubernetes version.

As of v1beta1, there is no "Kubernetes version" of a Cluster. There is only the Kubernetes version of the control plane, and of groups of machines (MachineDeployment/Pool).

ClusterClass adds the notion of a Kubernetes version of an entire cluster.

Lifecycle hooks also require ClusterClass.

Is this future work limited to ClusterClass clusters?

Reply:
What @Jont828 said is correct. This work will probably rely on the lifecycle hook work (and thus CC cluster only), but at this stage details are still TBD.